encoder/vaapiencoder_hevc.cpp

```diff
@@ -0,0 +1,1515 @@
+        headers.push_back(&m_sps);
+        headers.push_back(&m_pps);
+        uint8_t sync[] = {0, 0, 0, 1};
+        for (int i = 0; i < headers.size(); i++) {
+            m_headers.insert(m_headers.end(), sync, sync + N_ELEMENTS(sync));
+            appendHeaderWithEmulation(*headers[i]);
+        }
+    }
+
+    Encode_Status getCodecConfig(VideoEncOutputBuffer *outBuffer)
+    {
+        ASSERT((outBuffer && outBuffer->format == OUTPUT_CODEC_DATA) || outBuffer->format == OUTPUT_EVERYTHING);
+        if (outBuffer->bufferSize < m_headers.size())
+            return ENCODE_BUFFER_TOO_SMALL;
+        if (m_headers.empty())
+            return ENCODE_NO_REQUEST_DATA;
+        std::copy(m_headers.begin(), m_headers.end(), outBuffer->data);
+        outBuffer->dataSize = m_headers.size();
+        outBuffer->flag |= ENCODE_BUFFERFLAG_CODECCONFIG;
+        return ENCODE_SUCCESS;
+    }
+private:
+    BOOL bit_writer_write_vps (
+        BitWriter *bitwriter,
+        const VAEncSequenceParameterBufferHEVC* const seq,
+        VaapiProfile profile
+    )
+    {
+        BOOL vps_timing_info_present_flag = 0;
+
+        bit_writer_write_nal_header(bitwriter, VPS_NUT);
+
+        /* vps_video_parameter_set_id */
+        bit_writer_put_bits_uint32(bitwriter, 0, 4);
+        /* vps_base_layer_internal_flag */
+        bit_writer_put_bits_uint32(bitwriter, 1, 1);
+        /* vps_base_layer_available_flag */
+        bit_writer_put_bits_uint32(bitwriter, 1, 1);
+        /* vps_max_layers_minus1 */
+        bit_writer_put_bits_uint32(bitwriter, 0, 6);
+        /* vps_max_sub_layers_minus1 */
+        bit_writer_put_bits_uint32(bitwriter, 0, 3);
+        /* vps_temporal_id_nesting_flag */
+        bit_writer_put_bits_uint32(bitwriter, 1, 1);
+
+        /* vps_reserved_0xffff_16bits */
+        bit_writer_put_bits_uint32(bitwriter, 0xffff, 16);
+
+        profile_tier_level(bitwriter, 1, 0, seq);
+
+        /* vps_sub_layer_ordering_info_present_flag */
+        bit_writer_put_bits_uint32(bitwriter, 0, 1);
+
+        /*vps_max_dec_pic_buffering_minus1*/
+        bit_writer_put_ue(bitwriter, 6);
+
+        /*vps_max_num_reorder_pics*/
+        bit_writer_put_ue(bitwriter, 0);
+
+        /*vps_max_latency_increase_plus1*/
+        bit_writer_put_ue(bitwriter, 0);
+
+        /* vps_max_layer_id */
+        bit_writer_put_bits_uint32(bitwriter, 0, 6);
+
+        /* vps_num_layer_sets_minus1 */
+        bit_writer_put_ue(bitwriter, 0);
+
+        /* vps_timing_info_present_flag */
+        bit_writer_put_bits_uint32(bitwriter, vps_timing_info_present_flag, 1);
+
+        if (vps_timing_info_present_flag) {
+            /* TBD: add it later for BRC */
+        }
+
+        /* vps_extension_flag */
+        bit_writer_put_bits_uint32(bitwriter, 0, 1);
+
+        /* rbsp_trailing_bits */
+        bit_writer_write_trailing_bits(bitwriter);
+
+        return TRUE;
+    }
+
+    void st_ref_pic_set(BitWriter *bs, int stRpsIdx, int refIdx, const ShortRFS shortRFS)
```

Comment:
better use ShortRFS& instead ShortRFS.Else compile will make a copy for you. it harm performance